source/fem-dem/void_fraction.cc

```diff
@@ -0,0 +1,1141 @@
+          DoFTools::make_periodicity_constraints(
+            this->dof_handler,
+            id,
+            boundary_conditions.periodic_neighbor_id.at(id),
+            periodic_direction,
+            this->void_fraction_constraints);
+
+          has_periodic_boundaries = true;
+
+          // Get periodic offset if void fraction method is qcm or spm
+          if (this->void_fraction_parameters->mode ==
+                Parameters::VoidFractionMode::qcm ||
+              this->void_fraction_parameters->mode ==
+                Parameters::VoidFractionMode::spm)
+            {
+              periodic_offset = get_periodic_offset_distance(id);
+            }
+        }
+    }
+  void_fraction_constraints.close();
+
+
+  // Reinit system matrix
+
+  DynamicSparsityPattern dsp(locally_relevant_dofs);
+  DoFTools::make_sparsity_pattern(dof_handler,
+                                  dsp,
+                                  void_fraction_constraints,
+                                  false);
+
+  SparsityTools::distribute_sparsity_pattern(
+    dsp,
+    locally_owned_dofs,
+    this->triangulation->get_communicator(),
+    locally_relevant_dofs);
+
+  system_matrix_void_fraction.reinit(locally_owned_dofs,
+                                     locally_owned_dofs,
+                                     dsp,
+                                     this->triangulation->get_communicator());
+
+  if (has_periodic_boundaries)
+    LetheGridTools::vertices_cell_mapping_with_periodic_boundaries(
+      this->dof_handler, this->vertices_to_periodic_cell);
+}
+
+
+template <int dim>
+void
+VoidFractionBase<dim>::calculate_void_fraction(const double time)
+{
+  announce_string(this->pcout, "Void Fraction");
+
+  if (void_fraction_parameters->mode == Parameters::VoidFractionMode::function)
+    {
+      calculate_void_fraction_function(time);
+      // If its a function, no need to solve a linear system of equations so
+      // return.
+      return;
+    }
+  // The void fraction is established using a particle handler.
+  // A right-hand side and a linear system of equation are assembled and then
+  // solved. The resulting solution yields the nodal values of the void
+  // fraction.
+  if (void_fraction_parameters->mode == Parameters::VoidFractionMode::pcm)
+    {
+      calculate_void_fraction_particle_centered_method();
+    }
+  // particle_centered_method();
+  else if (void_fraction_parameters->mode == Parameters::VoidFractionMode::qcm)
+    {
+      calculate_void_fraction_quadrature_centered_method();
+    }
+
+  //  quadrature_centered_sphere_method(load_balance_step);
+  else if (void_fraction_parameters->mode == Parameters::VoidFractionMode::spm)
+    {
+      calculate_void_fraction_satellite_point_method();
+    }
+
+  solve_linear_system_and_update_solution(false);
+}
+
+template <int dim>
+void
+VoidFractionBase<dim>::calculate_void_fraction_function(const double time)
```

Comment:
Could time be passed by reference?

Reply:
yes but for a double that does not make a difference. The length of a pointer or a double is about the same